From e234058e1081cfc2080eecb911fcc8a30f32177b Mon Sep 17 00:00:00 2001
From: Austin Shafer <ashafer@nvidia.com>
Date: Sun, 16 Oct 2022 11:29:12 -0400
Subject: [PATCH 1/5] FreeBSD: pass address to kmem_free/pmap_unmapdev on 14.0
 and higher

This solves a compilation issue on CURRENT where we have a mismatch in
the first argument type for pmap_unmapdev and some kmalloc functions. As
of commit 7ae99f80b6661760c5de3edd330b279f04b092a2 the API changed, so
we should check the FreeBSD version and cast if needed.
---
 nvidia/src/nvidia/nv-freebsd.h | 9 ++++++++-
 nvidia/src/nvidia/nvidia_os.c  | 8 ++++++++
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/nvidia/src/nvidia/nv-freebsd.h b/nvidia/src/nvidia/nv-freebsd.h
index 517c888ccd3..2f081dce009 100644
--- a/nvidia/src/nvidia/nv-freebsd.h
+++ b/nvidia/src/nvidia/nv-freebsd.h
@@ -333,7 +333,14 @@ void   nv_unlock_api         (nv_state_t *);
         (obj) = NULL;                   \
     }
 
-#if __FreeBSD_version >= 1200083
+#if __FreeBSD_version >= 1400070
+#define NV_KMEM_ALLOC_CONTIG(size, flags, low, high, align, boundary, memattr) \
+    ((vm_offset_t)kmem_alloc_contig(size, flags, low, high, align, boundary, memattr))
+
+#define NV_KMEM_FREE(address, size) \
+    kmem_free((void *)address, size)
+
+#elif __FreeBSD_version >= 1200083
 #define NV_KMEM_ALLOC_CONTIG(size, flags, low, high, align, boundary, memattr) \
     kmem_alloc_contig(size, flags, low, high, align, boundary, memattr)
 
diff --git a/nvidia/src/nvidia/nvidia_os.c b/nvidia/src/nvidia/nvidia_os.c
index 6e0de980554..1a356bbfbf6 100644
--- a/nvidia/src/nvidia/nvidia_os.c
+++ b/nvidia/src/nvidia/nvidia_os.c
@@ -282,7 +282,15 @@ void NV_API_CALL os_unmap_kernel_space(
     NvU64 size
 )
 {
+    /*
+     * As of this FreeBSD version this function accepts a pointer value
+     * instead of casting it to a vm offset.
+     */
+#if __FreeBSD_version >= 1400070
+    pmap_unmapdev(address, size);
+#else
     pmap_unmapdev((vm_offset_t)address, size);
+#endif
 }
 
 void* NV_API_CALL os_map_user_space(
-- 
2.35.1

